dask/dataframe/operation.py

```diff
@@ -0,0 +1,596 @@
+from __future__ import annotations
+
+import itertools
+import operator
+from collections.abc import Mapping
+from functools import singledispatch
+from typing import Any
+
+import numpy as np
+
+from dask.base import tokenize
+from dask.dataframe.utils import make_meta
+from dask.highlevelgraph import HighLevelGraph
+from dask.operation import CollectionOperation, MemoizingVisitor, operations, regenerate
+from dask.utils import apply, is_arraylike
+
+
+class DataFrameOperation(CollectionOperation):
+
+    _name: str
+    _meta: Any
+    _divisions: tuple
+    _columns: set | None = None
+    applied_filters: list | None = None
+    projected_columns: set | None = None
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @name.setter
+    def name(self, value):
+        self._name = value
```

Comment:
The design of this feels to me like operations should be immutable, and creating one with different attributes should go through `regenerate()`. What do you think about that as a constraint, and changing `meta`, `name`, and `divisions` to being getters only?

My experience is that operating under constraints like that (and type annotations) is sometimes annoying, but pays off in the long run.

Reply:
Yes - We absolutely need a `CollectionOperation` object to be immutable, and so we should throw an error if the meta/divisions/etc setter is used.  However, we **do** need to allow these properties to be set for the specific case of `CompatFrameOperation` (for now).

Does this make sense? I will try to make this contract a bit clearer as I clean things up.